src/VisualStudio/Core/Def/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -192,160 +179,152 @@ private WatcherOperation(Kind kind, List<uint> cookies)
+            /// <param name="start">Start index (inclusive) of operations to combine.</param>
+            /// <param name="end">End index (inclusive) of operations to combine.</param>
+            public static WatcherOperation CombineRange(ImmutableSegmentedList<WatcherOperation> operations, int start, int end)
             {
-                if (other._kind == Kind.None)
-                {
-                    combined = this;
-                    return true;
-                }
-                else if (_kind == Kind.None)
-                {
-                    combined = other;
-                    return true;
-                }
-
-                switch (_kind)
-                {
-                    case Kind.WatchDirectory:
-                    case Kind.WatchFile:
-                        // Watching operations cannot be combined
-                        break;
-
-                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFile:
-                        combined = UnwatchFiles(ImmutableList.Create(_token, other._token));
-                        return true;
+                var firstOp = operations[start];
+                if (start == end)
+                    return firstOp;
 
-                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFiles:
-                        combined = UnwatchFiles(other._tokens.ToImmutableList().Insert(0, _token));
-                        return true;
+                using var _1 = ArrayBuilder<Context.RegularWatchedFile>.GetInstance(out var tokensBuilder);
+                using var _2 = ArrayBuilder<string>.GetInstance(out var fileNamesBuilder);
+                using var _3 = ArrayBuilder<uint>.GetInstance(out var cookiesBuilder);
 
-                    case Kind.UnwatchDirectories when other._kind == Kind.UnwatchDirectories:
-                        var cookies = new List<uint>(_cookies);
-                        cookies.AddRange(other._cookies);
-                        combined = UnwatchDirectories(cookies);
-                        return true;
-
-                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFile:
-                        combined = UnwatchFiles(_tokens.ToImmutableList().Add(other._token));
-                        return true;
-
-                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFiles:
-                        combined = UnwatchFiles(_tokens.ToImmutableList().AddRange(other._tokens));
-                        return true;
+                for (; start <= end; start++)
+                {
+                    var op = operations[start];
 
-                    default:
-                        break;
+                    switch (op._kind)
+                    {
+                        case Kind.WatchFiles:
+                            for (var i = 0; i < op._paths.Count; i++)
+                            {
+                                fileNamesBuilder.AddRange(op._paths[i]);
```

Comment:
Did you mean to do AddRange, because isn't that adding just a single one?

Reply:
Yeah, Add is what was intended here (I wanted to do AddRange but there isn't an ArrayBuilder.AddRange overload for OneOrMany)